src/common/Page.ts

```diff
@@ -2039,7 +2039,7 @@ export class Page extends EventEmitter {
       return false;
     }
 
-    return Promise.race([
```

Comment:
AFAIU, if there already are frames satisfying the predicate; their async execution must be faster (since we don't really wait for an event but execute a function on the page)

So, I wonder if we need to include that part in `Promise.race`. IMO, we can check it first, then if existing frames don't satisfy the predicate, we move on to waiting on events. WDYT?

It can be something like this:
```
const checkInExistingFrames = async () => {
      for (const frame of this.frames()) {
        if (await predicate(frame)) {
          return frame;
        }
      }
    }

    const foundFrame = await checkInExistingFrames();
    if (foundFrame) {
      return foundFrame;
    }

    return Promise.race([
      helper.waitForEvent(
        this._frameManager,
        FrameManagerEmittedEvents.FrameAttached,
        predicate,
        timeout,
        this._sessionClosePromise()
      ),
      helper.waitForEvent(
        this._frameManager,
        FrameManagerEmittedEvents.FrameNavigated,
        predicate,
        timeout,
        this._sessionClosePromise()
      ),
    ]);
```

This way it feel clearer what we do in the function, does it make sense?

Reply:
I have been thinking about that but since the predicate is async, I thought it would racey as we might get events while we are checking predicates for the existing frames. WDYT about that?

Reply:
Hihiiiim, that's true. Then, actually checking for existing frames and then waiting for events could be better for providing a deterministic behavior for that specific case:
* If there is already a frame on the page that satisfies the predicate, we'll first return that predicate
* if not, we'll wait for events

Otherwise, it can be indeterministic: i.e. Sometimes while there already is a frame on the page, before its predicate is executed, an event can be received and we return the event's frame. WDYT?

Reply:
> Otherwise, it can be indeterministic: i.e. Sometimes while there already is a frame on the page, before its predicate is executed, an event can be received and we return the event's frame. WDYT?

That is expected to be the same frame though (if you predicate is correct). Also, we don't guarantee any order in which you get frames so if the predicate is not correct, you get any first frame that is matching. We cannot really check predicates before setting up events listeners because that is not deterministic either.

Reply:
Oh, I now see which case that can be problematic, let's go with the race solution; thanks for the explanation! :)